### PR TITLE
Ftr: Gracefully shutdown

### LINKED ISSUE #239: Ftr:about Graceful Offline
**What would you like to be added**:
Problems encountered:

When the provider end does not get offline gracefully and does not receive the message with eventType of delete, the client will always keep the IP of the provider.

Solutions
1. The client should have the mechanism of kicking out the wrong IP: the client invoke should kick out the timeout IP, or reduce the priority of the timeout IP. Under the current code, the client can not remove the problematic IP. Even if you have a retry mechanism, now you have to wait for a timeout every time to retry. This is unacceptable.

2. Retry selection: all errors cannot be retried. Business exceptions should not be retried.
**Why is this needed**:

### PR DESCRIPTION
**Pull request by flycash**
**What this PR does**:
Gracefully shutdown. The steps are:
1. Destroy all registries;
2. wait for the new request for a short time
3. provider rejects new request and then wait for processing all requests;
4. consumer rejects sending new request and then wait for response
5. invoke custome shutdown callback
**Which issue(s) this PR fixes**:

Fixes #239 

**Special notes for your reviewer**:

Please check the order of releasing resources.

**Does this PR introduce a user-facing change?**:

```release-note
Now the application can shutdown gracefully
```

### CODE REVIEW

**Comment by flycash** [comment]
It seems that the syscall.SIGEMT cannot be found in CI.

**Comment by flycash** [comment]
I'm sorry that I had merge master and bring its change list into this PR

File: config/graceful_shutdown.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+	}
+	start := time.Now().UnixNano()
+	for time.Now().UnixNano()-start < timeout.Nanoseconds() && !shutdownConfig.RequestsFinished {
+		// sleep 10 ms and then we check it again
+		time.Sleep(10 * time.Millisecond)
+	}
+}
+
+func totalTimeout() time.Duration {
+	var providerShutdown time.Duration = 0
```
Comment (AlexStocks) [inline config/graceful_shutdown.go]: in Go, the default value is 0. so, delete "= 0" pls.

File: config/graceful_shutdown.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	if providerConfig != nil && providerConfig.ShutdownConfig != nil {
+		providerShutdown = providerConfig.ShutdownConfig.GetTimeout()
+	}
+
+	var consumerShutdown time.Duration = 0
```
Comment (AlexStocks) [inline config/graceful_shutdown.go]: delete "= 0" pls.

File: config/graceful_shutdown_config.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func (config *ShutdownConfig) Prefix() string {
+	return constant.ShutdownConfigPrefix
+}
+
+func (config *ShutdownConfig) GetTimeout() time.Duration {
+	result, err := time.ParseDuration(config.Timeout)
+	if err != nil {
+		logger.Errorf("The Timeout configuration is invalid: %s, and we will use the default value: %s",
```
Comment (AlexStocks) [inline config/graceful_shutdown_config.go]: you params num in this log is 2. However, the placeholder number is 2.

File: config/graceful_shutdown_config.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+			config.Timeout, defaultTimeout.String(), err)
+		return defaultTimeout
+	}
+	return result
+}
+
+func (config *ShutdownConfig) GetStepTimeout() time.Duration {
+	result, err := time.ParseDuration(config.StepTimeout)
+	if err != nil {
+		logger.Errorf("The StepTimeout configuration is invalid: %s, and we will use the default value: %s",
```
Comment (AlexStocks) [inline config/graceful_shutdown_config.go]: the same as above.

File: config/graceful_shutdown_test.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	})
+	extension.SetProtocol(constant.DUBBO, func() protocol.Protocol {
+		return &mockRegistryProtocol{}
+	})
+
+	extension.SetProtocol("mock", func() protocol.Protocol {
+		return &mockRegistryProtocol{}
+	})
+
+	protocolConfigs := make(map[interface{}]interface{})
```
Comment (AlexStocks) [inline config/graceful_shutdown_test.go]: pls set a default len for this map. maybe 16 is valid.

File: config/graceful_shutdown_test.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+	consumerConfig = &ConsumerConfig{
+		ProtocolConf: protocolConfigs,
+		ShutdownConfig: &ShutdownConfig{
+			Timeout:     "1",
+			StepTimeout: "1s",
+		},
+	}
+
+	providerProtocols := make(map[interface{}]interface{})
```
Comment (AlexStocks) [inline config/graceful_shutdown_test.go]: pls set a default len for this map. maybe 16 is valid.

File: filter/impl/graceful_shutdown_filter.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/config"
+	"github.com/apache/dubbo-go/filter"
+	"github.com/apache/dubbo-go/filter/common"
+	"github.com/apache/dubbo-go/protocol"
+)
+
+func init() {
+	var consumerFiler = &gracefulShutdownFilter{
+		activeCount:    0,
```
Comment (AlexStocks) [inline filter/impl/graceful_shutdown_filter.go]: If the init value is 0, you need not to set it.

File: filter/impl/graceful_shutdown_filter_test.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+ * limitations under the License.
+ */
+
+package impl
+
+import (
+	"net/url"
+	"testing"
+)
+import (
```
Comment (AlexStocks) [inline filter/impl/graceful_shutdown_filter_test.go]: add a blank bf this line.

File: filter/impl/graceful_shutdown_filter_test.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/stretchr/testify/assert"
+)
+import (
```
Comment (AlexStocks) [inline filter/impl/graceful_shutdown_filter_test.go]: the same. add a blank line pls.

File: config/graceful_shutdown_config_test.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+
+package config
+
+import (
+	"testing"
+	"time"
+)
+import (
```
Comment (AlexStocks) [inline config/graceful_shutdown_config_test.go]: add a blank line bf this line.

File: config/graceful_shutdown.go [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+
+	signal.Notify(signals, os.Interrupt, os.Kill, syscall.SIGKILL, syscall.SIGSTOP,
+		syscall.SIGHUP, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGILL, syscall.SIGTRAP,
+		syscall.SIGABRT, syscall.SIGSYS,
+	)
+
+	go func() {
+		select {
+		case sig := <-signals:
+			logger.Infof("get signal %s, application will shutdown.", sig.String())
```
Comment (AlexStocks) [inline config/graceful_shutdown.go]: sig.String() -> sig.

File: config/graceful_shutdown.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+func BeforeShutdown() {
+
+	destroyAllRegistries()
+	// waiting for a short time so that the clients have enough time to get the notification that server shutdowns
+	// The value of configuration depends on how long the clients will get notification.
+	waitAndAcceptNewRequests()
+
+	time.Sleep(1 * time.Minute)
```
Comment (AlexStocks) [inline config/graceful_shutdown.go]: one minute is not a short time. @hxmhlt @fangyincheng , how long is enough?

File: config/graceful_shutdown.go [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+	waitingProcessedTimeout(consumerConfig.ShutdownConfig)
+}
+
+func waitingProcessedTimeout(shutdownConfig *ShutdownConfig) {
+	timeout := shutdownConfig.GetStepTimeout()
+	if timeout <= 0 {
+		return
+	}
+	start := time.Now().UnixNano()
+	for time.Now().UnixNano()-start < timeout.Nanoseconds() && !shutdownConfig.RequestsFinished {
```
Comment (AlexStocks) [inline config/graceful_shutdown.go]: pls use time.Time.After or time.Time.Before instead.

File: config/graceful_shutdown_config.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+import (
+	"time"
+)
+import (
```
Comment (AlexStocks) [inline config/graceful_shutdown_config.go]: add a blank line before this line.

File: config/graceful_shutdown_test.go [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+import (
+	"testing"
+)
+import (
```
Comment (AlexStocks) [inline config/graceful_shutdown_test.go]: add a blank line before this import.

File: config/graceful_shutdown.go (hunk 12 above)
Comment (pantianying) [reply config/graceful_shutdown.go]: I think 10 seconds is enough

Comment (pantianying) [inline config/graceful_shutdown.go]: Is this the same as waitAndAcceptNewRequests(),Is it repeated？

Comment (flycash) [reply config/graceful_shutdown.go]: Sorry, it's for tests, I forgot to remove it.

Comment (flycash) [reply config/graceful_shutdown.go]: It was removed. It's my testing data... the waitAndAcceptNewRequests() will do the waiting, and the time is ShutdownConfig.StepTimeout, and the default value is 10s

**Comment by flycash** [comment]
All comments had been fixed and I've tested it end to end again.

**Comment by AlexStocks** [comment]
LGTM

**Comment by pantianying** [comment]
LGTM

**Pull request merged**